tools/castle-editor/code/framedesign.pas

```diff
@@ -1760,9 +1755,9 @@ procedure TDesignFrame.CastleControlUpdate(Sender: TObject);
 
   if InternalCastleDesignInvalidate then
   begin
-    ModifiedOutsideObjectInspector;
     UpdateDesign;
     //WritelnWarning('CGE needed to explicitly tell editor to refresh hierarchy');
+    ModifiedOutsideObjectInspector('Internal update'); //WARNING: I don't really know when this happens and how it should behave
```

Comment:
`InternalCastleDesignInvalidate` means that something in CGE changed hierarchy. 

Right now it happens when `ExposeTransforms` is set and in effect new `TCastleTransform` is created and added as child of some existing `TCastleTransform`. In general it will happen always when something in CGE needs to change UI/transform hierarchy but we did not handle this change (by calling `UpdateDesign` explicitly) elsewhere in editor (e.g. `InternalCastleDesignInvalidate` is also set when creating new component, but in this case it is handled by `UpdateDesign` earlier).

Calling it an "Internal update" is not accurate, something changed in response to user actions and the change is visible, it's not really "internal" :)

So just make it `ModifiedOutsideObjectInspector('');`

Reply:
I see...

Here let's make one more decision. Now Undo recorded with `Comment := ''` actually transforms into:

```Pascal
if UndoHistory[UndoI].Comment = '' then
begin
  if UndoHistory[UndoI].Selected = '' then
    Result := 'Modify global variable'
  else
    Result := 'Modify ' + UndoHistory[UndoI].Selected;
end
```

I.e. in this case we'll get 'Modify global variable' which is not actually correct. Maybe we should change this feature in core. E.g. in this case let's the menu item caption be just `'Undo'` without any specifics? WDYT?

Reply:
Oh, I see. Yes, I expected `UndoComment=''` to make just `'Undo'` menu caption without any specifics.

The fix should be a little more involved than just changing `'Modify global variable'` case. Because there may be something selected at this moment, and we still want to just show `'Undo'` menu caption.

I guess we have a problem with competing ideas how to treat `UndoComment=''`. E.g. I see `AdjustUserInterfaceAnchorsToKeepRect` also passes `UndoComment = ''`, and in this case it makes sense that it results in `'Modify ' + Selected`.

Proposed solution: When things are unclear, introduce an enumerated type to make it explicit and clear :) Like this:

```pascal
TUndoType = (
  // This undo has no description (UndoComment is ignored and should be left empty).
  utNoDescription,
  // This undo has description in UndoComment.
  utDescription,
  // This undo should have description like "Modify SelectedComponentName" (UndoComment is ignored and should be left empty).
  utModifiedSelected,
  // No undo is recorded, defer recording it to OnRelease. (UndoComment is ignored and should be left empty).
  utDeferToOnRelease
);
```

And now:

- `ModifiedOutsideObjectInspector` takes `const UndoType: TUndoType` as 1st parameter

- `ModifiedOutsideObjectInspector` no longed needs `UndoOnRelease` boolean, `utDeferToOnRelease` signals this purpose.

- We should explicitly forbid "Modify global variable", because that should not be really possible (you cannot do this in editor, and you could, we could not be able to undo it -- as we only save/restore the design).

So add to `ModifiedOutsideObjectInspector` 

```
{ When UndoType is utDescription, then non-empty UndoComment is required.
  When UndoType is something else, then UndoComment is ignored, and we require it to be empty for now. }
Assert((UndoComment <> '') = (UndoType = utDescription));

{ Using utModifiedSelected is only allowed when something is selected. }
Assert((UndoType <> utModifiedSelected) or (SelectedC <> nil));
```

Reply:
After 2 days of thinking over and struggling to find a good solution (without asserts), I've came up with a (significantly) different idea. The problem was that we treated `''` as a "magic" comment that implied "low priority comment". And that "magic" actually harmed us in the way described above :) In the new approach I explicitly added "UndoCommentPriority" which specifies how high the priority of the given comment is.

As a by-product I also significantly improved the undo comments themselves, and made them as informative as it seems to be possible.